Improve consistency of cache key generation with args/kwargs

nbecker-yc opened the pull request:
### Overview
Ticket: https://app.asana.com/1/1144271657183267/project/1203633948717779/task/1209820190967177?focus=true

### How to test
- [x] Review code/run tests

dlim-yc reviewed (commented):
Started by reviewing the tests, and came up with this edge case where the cache key generation could be optimized even further:

```
        inc_1 = incrementer.instance_increment_by_with_kwargs(1, None)
        inc_2 = incrementer.instance_increment_by_with_kwargs(1, useless_kwarg=None)
        inc_3 = incrementer.instance_increment_by_with_kwargs(1)
```

All three of these could technically be cached at the same key, since the default value of the kwarg is `None`. I believe you covered for 1 == 2, but not 3

Does this seem feasible?

nbecker-yc commented:
> Started by reviewing the tests, and came up with this edge case where the cache key generation could be optimized even further:
> 
> ```
>         inc_1 = incrementer.instance_increment_by_with_kwargs(1, None)
>         inc_2 = incrementer.instance_increment_by_with_kwargs(1, useless_kwarg=None)
>         inc_3 = incrementer.instance_increment_by_with_kwargs(1)
> ```
> 
> All three of these could technically be cached at the same key, since the default value of the kwarg is `None`. I believe you covered for 1 == 2, but not 3
> 
> Does this seem feasible?

@dlim-yc great catch. just fixed this and added 3 additional tests (one for each type of decorator).

dlim-yc reviewed (approved):
Looks great! Pretty elegant solution

nbecker-yc commented on cache_helper/decorators.py:
```diff
@@ -15,14 +17,32 @@
+    hashed_function_cache_key = utils.get_hashed_cache_key(function_cache_key)
+    return hashed_function_cache_key, function_cache_key
+
+
 def cached(timeout):
     def _cached(func):
         func_name = utils.get_function_name(func)
+        func_signature = signature(func)
 
         @wraps(func)
         def wrapper(*args, **kwargs):
-            function_cache_key = utils.get_function_cache_key(func_name, args, kwargs)
-            cache_key = utils.get_hashed_cache_key(function_cache_key)
+            cache_key, function_cache_key = _get_function_cache_keys(func_name, func_signature, args, kwargs)
```
yeah definitely agreed. I left the names as they were to minimize changes but agree that's cleaner. I'll update accordingly.

nbecker-yc replied:
done!
